inode.c

```diff
@@ -110,6 +110,77 @@ struct inode *simplefs_iget(struct super_block *sb, unsigned long ino)
     return ERR_PTR(ret);
 }
 
+static int __file_lookup(struct inode *dir,
+                         struct dentry *dentry,
+                         int *ei,
+                         int *bi,
+                         int *fi)
+{
+    struct super_block *sb = dir->i_sb;
+    struct simplefs_inode_info *ci_dir = SIMPLEFS_INODE(dir);
+    struct buffer_head *bh = NULL, *bh2 = NULL;
+    struct simplefs_file_ei_block *eblock = NULL;
+    struct simplefs_dir_block *dblock = NULL;
+    struct simplefs_file *f = NULL;
+    int _ei, _bi, _fi, ret = 0;
+    int idx_ei, idx_bi;
+    uint32_t hash_code;
+    /* Check filename length */
+
+    /* Read the directory block on disk */
+    bh = sb_bread(sb, ci_dir->ei_block);
+    if (!bh)
+        return -EIO;
+    eblock = (struct simplefs_file_ei_block *) bh->b_data;
+    hash_code = simplefs_hash(dentry) %
+                (SIMPLEFS_MAX_EXTENTS * SIMPLEFS_MAX_BLOCKS_PER_EXTENT);
+    _ei = hash_code / SIMPLEFS_MAX_BLOCKS_PER_EXTENT;
+    _bi = hash_code % SIMPLEFS_MAX_BLOCKS_PER_EXTENT;
+
+    for (idx_ei = 0; idx_ei < SIMPLEFS_MAX_EXTENTS; _ei++, idx_ei++) {
+        if (unlikely(_ei >= SIMPLEFS_MAX_EXTENTS))
+            _ei %= SIMPLEFS_MAX_EXTENTS;
```

Comment:
The wraparound search (`_ei %= SIMPLEFS_MAX_EXTENTS`) handles collisions, but the worst case is still O(n) if many files hash to the same bucket.

Consider documenting expected collision rates with your benchmark files.